src/interpreter/evaluator.ts

```diff
@@ -0,0 +1,128 @@
+import {
+  type EvaluatorMapping,
+  type EvaluatorState,
+  type Result,
+  type Value
+} from './types/evaluator';
+import {
+  type AssignInstr,
+  type BinaryOperationInstr,
+  type CallInstr,
+  type DoneInstr,
+  type EnterProgramInstr,
+  type GotoInstr,
+  type JumpOnFalseInstr,
+  type LoadConstantInstr,
+  type LoadFunctionInstr,
+  type LoadSymbolInstr,
+  type TeardownInstr
+} from './types/instruction';
+import {
+  convertToAddress,
+  convertToPredicate,
+  evaluateBinaryExpression,
+  isTrue,
+  typeCheckBinaryOperation
+} from './evaluatorUtils';
+import { type CompilerState } from './types/virtualMachine';
+import { Stack } from '../utils/stack';
+import { type Program } from '../ast/types';
+import { compileProgram } from './virtualMachine';
+
+/**
+ * Evaluates the abstract syntax tree using a virtual machine evaluator &
+ * returns the result of evaluation asynchronously.
+ *
+ * @param ast The abstract syntax tree to evaluate.
+ */
+export const evaluate = async (ast: Program): Promise<Result> => {
+  return await new Promise(
+    (
+      resolve: (value: Result | PromiseLike<Result>) => void,
+      _reject: (reason?: any) => void
+    ) => {
+      try {
+        const compilation = compileProgram(ast);
+        const value = interpret(compilation);
+        resolve({ status: 'finished', value });
+      } catch (err) {
+        resolve({ status: 'error' });
+      }
+    }
+  );
+};
+
+export const interpret = (compilation: CompilerState): Value => {
+  const stash = new Stack<Value>();
+  const state: EvaluatorState = {
+    memory: compilation.memory,
```

Comment:
Yea, it seems quite weird that the program memory is tied with the virtual machine code because it is now inherently stateful when conventionally, it should be stateless.